Sometimes, local files do not load

pvollan opened the pull request:
#### 5a1ce4d09d78d43e1f6799091475f131a370f0d7
<pre>
Sometimes, local files do not load
<a href="https://bugs.webkit.org/show_bug.cgi?id=290474">https://bugs.webkit.org/show_bug.cgi?id=290474</a>

Reviewed by Ben Nham.

When loading local files, the system is providing the UI process with a sandbox extension to the file or its parent
directory. If the system is just providing the sandbox extension to the file itself, we will fail to create a
sandbox extension in the UI process for the WebContent process. This patch addresses this by adding a fallback to
WebPageProxy::maybeInitializeSandboxExtensionHandle. If we fail to create a sandbox extension for the parent
directory, we try to create a sandbox extension for the file itself. This patch also switches to a preferred
sandbox extension method in WebPageProxy::createSandboxExtensionsIfNeeded, which is based on audit token.

* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::createSandboxExtensionsIfNeeded):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::maybeInitializeSandboxExtensionHandle):

Canonical link: <a href="https://commits.webkit.org/292762@main">https://commits.webkit.org/292762@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/946810d9335f7234768b1ad42b41e08b632e300a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96983 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16597 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6799 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102059 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47504 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16893 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25046 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73880 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31088 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99986 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12737 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87724 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54216 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12493 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46831 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82573 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5641 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104080 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24051 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17553 "Found 1 new test failure: imported/w3c/web-platform-tests/workers/abrupt-completion.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82928 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24304 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83813 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82323 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27005 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4553 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17554 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15654 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24013 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29168 "Failed to build and analyze WebKit") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23840 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27152 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25413 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham reviewed (approved):
Maybe the "try `SandboxExtension::createHandleForReadByAuditToken` but then fall back to `SandboxExtension::createHandle`" logic can be abstracted?

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -1974,12 +1974,38 @@ void WebPageProxy::maybeInitializeSandboxExtensionHandle(WebProcessProxy& proces
+    if (auto token = process.protectedConnection()->getAuditToken()) {
+        if (auto handle = SandboxExtension::createHandleForReadByAuditToken(fullPath, *token)) {
+            sandboxExtensionHandle = WTFMove(*handle);
+            createdExtension = true;
```
Boolean variables in WebKit need a prefix, per WebKit coding style.

pvollan replied:
Ah, good catch! This boolean will be removed when I do the abstraction in the follow-up patch coming shortly.

Thanks for reviewing!

pvollan committed:
Sometimes, local files do not load
https://bugs.webkit.org/show_bug.cgi?id=290474

Reviewed by Ben Nham.

When loading local files, the system is providing the UI process with a sandbox extension to the file or its parent
directory. If the system is just providing the sandbox extension to the file itself, we will fail to create a
sandbox extension in the UI process for the WebContent process. This patch addresses this by adding a fallback to
WebPageProxy::maybeInitializeSandboxExtensionHandle. If we fail to create a sandbox extension for the parent
directory, we try to create a sandbox extension for the file itself. This patch also switches to a preferred
sandbox extension method in WebPageProxy::createSandboxExtensionsIfNeeded, which is based on audit token.

* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::createSandboxExtensionsIfNeeded):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::maybeInitializeSandboxExtensionHandle):

Canonical link: https://commits.webkit.org/292762@main
